File: sdk/ts/src/marketplace.ts
import fetch from "node-fetch";

export interface ListingParams {
  name: string;
  priceAsset: "XLM" | "USDC";
  pricePerCall: string;
  endpoint: string;
  schema: Record<string, string>;
}

export interface Listing extends ListingParams {
  id: string;
  active: boolean;
  sellerPubkey: string;
  createdAt: string;
  updatedAt: string;
}

export interface MarketplaceOptions {
  backendUrl?: string;
}

export class Marketplace {
  private backendUrl: string;

  constructor(opts: MarketplaceOptions = {}) {
    this.backendUrl = opts.backendUrl || "http://localhost:4000";
  }

  async registerListing(
    params: ListingParams & { sellerPubkey: string },
  ): Promise<Listing> {
    const pricePerCallStroops = Math.floor(
      parseFloat(params.pricePerCall) * 10_000_000,
    );
    const response = await fetch(
      `${this.backendUrl}/api/v1/marketplace/listings`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          name: params.name,
          sellerPubkey: params.sellerPubkey,
          endpoint: params.endpoint,
          pricePerCall: pricePerCallStroops,
          asset: params.priceAsset,
          schema: params.schema,
        }),
      },
    );

    if (!response.ok) {
      const errorData = (await response.json()) as { error: string };
      throw new Error(errorData.error || "Failed to register listing");
    }

    const data = (await response.json()) as { listing: any };
    return {
      id: data.listing.id,
      name: data.listing.name,
      priceAsset: data.listing.asset,
      pricePerCall: (data.listing.pricePerCall / 10_000_000).toString(),
      endpoint: data.listing.endpoint,
      schema: data.listing.schema,
      active: data.listing.active,
      sellerPubkey: data.listing.sellerPubkey,
      createdAt: data.listing.createdAt,
      updatedAt: data.listing.updatedAt,
    };
  }

  async getListings(filters?: {
    name?: string;
    sellerPubkey?: string;
    asset?: string;
    activeOnly?: boolean;
    limit?: number;
    offset?: number;
  }): Promise<{
    listings: Listing[];
    total: number;
    limit: number;
    offset: number;
  }> {
    const queryParams = new URLSearchParams();
    if (filters) {
      if (filters.name) queryParams.append("name", filters.name);
      if (filters.sellerPubkey)
        queryParams.append("seller", filters.sellerPubkey);
      if (filters.asset) queryParams.append("asset", filters.asset);
      if (filters.activeOnly !== undefined)
        queryParams.append("activeOnly", filters.activeOnly.toString());
      if (filters.limit) queryParams.append("limit", filters.limit.toString());
      if (filters.offset)
        queryParams.append("offset", filters.offset.toString());
    }

    const response = await fetch(
      `${this.backendUrl}/api/v1/marketplace/listings?${queryParams.toString()}`,
    );
    if (!response.ok) {
      const errorData = (await response.json()) as { error: string };
      throw new Error(errorData.error || "Failed to get listings");
    }

    const data = (await response.json()) as {
      listings: any[];
      total: number;
      limit: number;
      offset: number;
    };
    return {
      listings: data.listings.map((listing) => ({
        id: listing.id,
        name: listing.name,
        priceAsset: listing.asset,
        pricePerCall: (listing.pricePerCall / 10_000_000).toString(),
        endpoint: listing.endpoint,
        schema: listing.schema,
        active: listing.active,
        sellerPubkey: listing.sellerPubkey,
        createdAt: listing.createdAt,
        updatedAt: listing.updatedAt,
      })),
      total: data.total,
      limit: data.limit,
      offset: data.offset,
    };
  }
}

File: sdk/rust/src/marketplace.rs
use serde::{Deserialize, Serialize};
use reqwest::Client;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListingParams {
    pub name: String,
    pub price_asset: String,
    pub price_per_call: String,
    pub endpoint: String,
    pub schema: std::collections::HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Listing {
    pub id: String,
    pub name: String,
    pub price_asset: String,
    pub price_per_call: String,
    pub endpoint: String,
    pub schema: std::collections::HashMap<String, String>,
    pub active: bool,
    pub seller_pubkey: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone)]
pub struct MarketplaceOptions {
    pub backend_url: Option<String>,
}

pub struct Marketplace {
    backend_url: String,
    client: Client,
}

impl Marketplace {
    pub fn new(opts: MarketplaceOptions) -> Self {
        Marketplace {
            backend_url: opts.backend_url.unwrap_or_else(|| "http://localhost:4000".to_string()),
            client: Client::new(),
        }
    }

    pub async fn register_listing(&self, params: ListingParams, seller_pubkey: String) -> Result<Listing, String> {
        let price_per_call_stroops = (params.price_per_call.parse::<f64>().unwrap() * 10_000_000.0) as i64;

        let request_body = serde_json::json!({
            "name": params.name,
            "sellerPubkey": seller_pubkey,
            "endpoint": params.endpoint,
            "pricePerCall": price_per_call_stroops,
            "asset": params.price_asset,
            "schema": params.schema,
        });

        let response = self.client
            .post(format!("{}/api/v1/marketplace/listings", self.backend_url))
            .json(&request_body)
            .send()
            .await
            .map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(error_text);
        }

        let data: serde_json::Value = response.json().await.map_err(|e| e.to_string())?;
        let listing = data.get("listing").ok_or("listing not found in response")?;

        Ok(Listing {
            id: listing.get("id").unwrap().as_str().unwrap().to_string(),
            name: listing.get("name").unwrap().as_str().unwrap().to_string(),
            price_asset: listing.get("asset").unwrap().as_str().unwrap().to_string(),
            price_per_call: (listing.get("pricePerCall").unwrap().as_i64().unwrap() as f64 / 10_000_000.0).to_string(),
            endpoint: listing.get("endpoint").unwrap().as_str().unwrap().to_string(),
            schema: serde_json::from_value(listing.get("schema").unwrap().clone()).unwrap(),
            active: listing.get("active").unwrap().as_bool().unwrap(),
            seller_pubkey: listing.get("sellerPubkey").unwrap().as_str().unwrap().to_string(),
            created_at: listing.get("createdAt").unwrap().as_str().unwrap().to_string(),
            updated_at: listing.get("updatedAt").unwrap().as_str().unwrap().to_string(),
        })
    }
}
